src/probe/probe.c

```diff
@@ -15,42 +31,1072 @@
+	buffer->avail = 0;
+
+	return 0;
+}
+
+/**
+ * \brief Request DMA and initialize DMA for probes with correct alignment,
+ *	  size and specific channel.
+ * \param[out] probe DMA.
+ * \param[in] direction.
+ * \return 0 on success, error code otherwise.
+ */
+static int probe_dma_init(struct probe_dma_ext *dma, uint32_t direction)
+{
+	struct dma_sg_config config;
+	uint32_t elem_addr, addr_align;
+	const uint32_t elem_size = sizeof(uint64_t) * DMA_ELEM_SIZE;
+	const uint32_t elem_num = PROBE_BUFFER_LOCAL_SIZE / elem_size;
+	int err = 0;
+
+	/* request DMA in the dir LMEM->HMEM with shared access */
+	dma->dc.dmac = dma_get(direction, 0, DMA_DEV_HOST,
+			       DMA_ACCESS_SHARED);
+	if (!dma->dc.dmac) {
+		trace_probe_error("probe_dma_init() error: dma->dc.dmac = NULL");
+		return -ENODEV;
+	}
+
+	/* get required address alignment for dma buffer */
+	err = dma_get_attribute(dma->dc.dmac, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+				&addr_align);
+	if (err < 0)
+		return err;
+
+	/* initialize dma buffer */
+	err = probe_dma_buffer_init(&dma->dmapb, PROBE_BUFFER_LOCAL_SIZE, addr_align);
+	if (err < 0)
+		return err;
+
+	err = dma_copy_set_stream_tag(&dma->dc, dma->stream_tag);
+	if (err < 0)
+		return err;
+
+	elem_addr = (uint32_t)dma->dmapb.addr;
+
+	config.direction = direction;
+	config.src_width = sizeof(uint32_t);
+	config.dest_width = sizeof(uint32_t);
+	config.cyclic = 0;
+
+	err = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+	if (err < 0)
+		return err;
+
+	err = dma_set_config(dma->dc.chan, &config);
+	if (err < 0)
+		return err;
+
+	dma_sg_free(&config.elem_array);
+
+
+	return 0;
+}
+
+/**
+ * \brief Stop, deinit and free DMA and buffer used by probes.
+ * \param[out] probe DMA.
+ * \return 0 on success, error code otherwise.
+ */
+static int probe_dma_deinit(struct probe_dma_ext *dma)
+{
+	int err = 0;
+
+	err = dma_stop(dma->dc.chan);
+	if (err < 0) {
+		trace_probe_error("probe_dma_deinit() error: dma_stop() failed");
+		return err;
+	}
+
+	dma_channel_put(dma->dc.chan);
+	dma_put(dma->dc.dmac);
+
+	rfree((void *)dma->dmapb.addr);
+	dma->dmapb.addr = 0;
+
+	dma->stream_tag = PROBE_DMA_INVALID;
+
+	return 0;
+}
+
+/*
+ * \brief Probe task for extraction.
+ *
+ * Copy extraction probes data to host if available.
+ * Return err if dma copy failed.
+ */
+static enum task_state probe_task(void *data)
+{
+	struct probe_pdata *_probe = probe_get();
+	int err;
+
+	if (_probe->ext_dma.dmapb.avail > 0)
+		err = dma_copy_to_host_nowait(&_probe->ext_dma.dc,
+					      &_probe->ext_dma.config, 0,
+					       (void *)_probe->ext_dma.dmapb.r_ptr,
+					       _probe->ext_dma.dmapb.avail);
+	else
+		return SOF_TASK_STATE_RESCHEDULE;
+
+	if (err < 0) {
+		trace_probe_error("probe_task() error: dma_copy_to_host_nowait() failed.");
+		return err;
+	}
+
+	_probe->ext_dma.dmapb.r_ptr = _probe->ext_dma.dmapb.w_ptr;
+	_probe->ext_dma.dmapb.avail = 0;
```

Comment:
A comment stating that you are emptying the buffer once you have transferred it would be nice.

Reply:
done